<a href="https://colab.research.google.com/github/estebanramirezcantillo/Esteban-Ramirez---clase-AdN/blob/main/RETO2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Reto 2 - Esteban Ramirez

#1 datos

In [ ]:
data1 = {
    "Variable": ["Sales_Rep", "Business", "Age", "Female", "Years", "College", "Certificates", "Feedback", "Salary", "NPS"],
    "Descripción": [
        "Identificador único de cada representante de ventas.",
        "Tipo de negocio en el que trabaja el representante (Hardware o Software).",
        "Edad del representante de ventas.",
        "Indicador de género (1 si es mujer, 0 si es hombre).",
        "Años de experiencia en la empresa.",
        "Indica si el representante tiene estudios universitarios (Yes/No).",
        "Número de certificaciones obtenidas.",
        "Puntuación de retroalimentación de desempeño.",
        "Salario anual del representante en dólares.",
        "Net Promoter Score, mide la lealtad del cliente en una escala de 1 a 10."
    ]
}

df_descripcion = pd.DataFrame(data1)
pd.set_option('display.max_colwidth', None)
print(df_descripcion.to_string(index=False))


    Variable                                                               Descripción
   Sales_Rep                      Identificador único de cada representante de ventas.
    Business Tipo de negocio en el que trabaja el representante (Hardware o Software).
         Age                                         Edad del representante de ventas.
      Female                      Indicador de género (1 si es mujer, 0 si es hombre).
       Years                                        Años de experiencia en la empresa.
     College        Indica si el representante tiene estudios universitarios (Yes/No).
Certificates                                      Número de certificaciones obtenidas.
    Feedback                             Puntuación de retroalimentación de desempeño.
      Salary                               Salario anual del representante en dólares.
         NPS  Net Promoter Score, mide la lealtad del cliente en una escala de 1 a 10.


###cargar librerias y datos

In [ ]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
nxl=('/2. TechSalesReport_Int_L (1).xlsx')

In [ ]:
XDB=pd.read_excel(nxl,sheet_name=0)
XDB['College']= XDB['College'].replace({'Yes':1,'No':0})

XDB['Business']= XDB['Business'].replace({'Hardware':0,'Software':1})
# Variables a Utilizar
XD=XDB[['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']]
yd=XDB[['College']]

<ipython-input-45-de12479939e4>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['College']= XDB['College'].replace({'Yes':1,'No':0})
<ipython-input-45-de12479939e4>:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['Business']= XDB['Business'].replace({'Hardware':0,'Software':1})


##2. Modelo - Arbol de Decision

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)
# librerias arbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data
vs=['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']
dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('Reto2Arbol.png')


True

##3. Caracterización de la base de datos

In [ ]:

Total_Personas = len(XDB)
Total_Inscritos = VP + FN
Total_NoInscritos = VN + FP

print("\n===== 📊 Resumen de Datos =====")
print(f"👥 Total de personas analizadas: {Total_Personas}")
print(f"✅ Total de personas inscritas: {Total_Inscritos}")
print(f"❌ Total de personas no inscritas: {Total_NoInscritos}")



===== 📊 Resumen de Datos =====
👥 Total de personas analizadas: 21990
✅ Total de personas inscritas: 17520
❌ Total de personas no inscritas: 4470


## 4. Nodo Puro
No hay nodo puro.

##.5 Pronostico

In [ ]:
# Cargar los datos (sheet 1)
XDB_pred = pd.read_excel(nxl, sheet_name=1)
XDB_pred['Business'] = pd.factorize(XDB_pred['Business'])[0]
XDB_pred['Business']= XDB_pred['Business'].replace({'Hardware':0,'Software':1})
XD_pred = XDB_pred[['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']]
# Realizar las predicciones
yd_pred = mar.predict(XD_pred)
print("Las predicciones son:\n", yd_pred)

# Crear el informe con las predicciones
XDB_pred['College_Predicted'] = yd_pred
print(XDB_pred.head(10))



Las predicciones son:
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
   Sales_Rep  Business  Age  Female  Years  ... Certificates  Feedback  Salary  NPS  College_Predicted
0          1         0   63       0      2  ...            5      3.49  133000    8                  1
1          2         1   38       1      2  ...            4      3.33   80400    6                  1
2          3         1   47       0      1  ...            2      2.12   57000    7                  1
3          4         1   25       0      1  ...            2      2.89   79000    7                  1
4          5         0   49       0     10  ...            4      2.52  127000    6                  1
5          6         1   48       0      1  ...            1      3.35   62000    6                  1
6          7         1   25       1      1  ...            6      1.16   74000    7                  1
7          8         0   39       0      1  ...            2      2.12  112000    5                  1
8          9    

##6.Desempeño modelo

In [ ]:
ydp=mar.predict(XD)
cm = confusion_matrix(yd.values.ravel(), ydp)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
ex = (VN + VP) / (VN + VP + FN + FP)  # Exactitud
TE = (FN + FP) / (VN + VP + FN + FP)  # Tasa de error
Sen = VP / (VP + FN) if (VP + FN) != 0 else 0  # Sensibilidad (Recall)
Esp = VN / (VN + FP) if (VN + FP) != 0 else 0  # Especificidad
preP = VP / (VP + FP) if (VP + FP) != 0 else 0  # Precisión de positivos
preN = VN / (VN + FN) if (VN + FN) != 0 else 0  # Precisión de negativos

# Imprimir resultados en un formato más claro
print("\n===== Evaluación del Modelo =====")
print("🔹 Matriz de Confusión:")
print(cm)

print("\n🔹 Métricas de Desempeño:")
print(f"✅ Exactitud: {ex:.4f}")
print(f"❌ Tasa de Error: {TE:.4f}")
print(f"📊 Sensibilidad (Recall): {Sen:.4f}")
print(f"⚖️  Especificidad: {Esp:.4f}")
print(f"🎯 Precisión Positivos: {preP:.4f}")


===== Evaluación del Modelo =====
🔹 Matriz de Confusión:
[[  114  4356]
 [   69 17451]]

🔹 Métricas de Desempeño:
✅ Exactitud: 0.7988
❌ Tasa de Error: 0.2012
📊 Sensibilidad (Recall): 0.9961
⚖️  Especificidad: 0.0255
🎯 Precisión Positivos: 0.8002


## Informe en excel

In [ ]:
# Crear el informe con la preaprobacion
dfp=pd.DataFrame(np.column_stack((XD_pred,yd_pred)))
dfp.columns=[['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS','College_Predicted']]
dfp.head(15)
dfp.to_excel('InformePRED.xlsx')
print(dfp.head(15))

   Business   Age Female Years Certificates Feedback    Salary   NPS College_Predicted
0       0.0  63.0    0.0   2.0          5.0     3.49  133000.0   8.0               1.0
1       1.0  38.0    1.0   2.0          4.0     3.33   80400.0   6.0               1.0
2       1.0  47.0    0.0   1.0          2.0     2.12   57000.0   7.0               1.0
3       1.0  25.0    0.0   1.0          2.0     2.89   79000.0   7.0               1.0
4       0.0  49.0    0.0  10.0          4.0     2.52  127000.0   6.0               1.0
5       1.0  48.0    0.0   1.0          1.0     3.35   62000.0   6.0               1.0
6       1.0  25.0    1.0   1.0          6.0     1.16   74000.0   7.0               1.0
7       0.0  39.0    0.0   1.0          2.0     2.12  112000.0   5.0               1.0
8       1.0  29.0    0.0   1.0          5.0     2.74   78000.0   8.0               1.0
9       1.0  45.0    1.0   3.0          3.0     2.90   74000.0   7.0               1.0
10      0.0  35.0    0.0   6.0          1.0

## Conclusiones
El modelo de árbol de decisión ha predicho que todas las personas en la muestra serán admitidas en el college (College_Predicted = 1.0). En general, el modelo sugiere que casi cualquier persona en la muestra cumple con los requisitos para ser admitida, lo que puede indicar un sesgo en los datos de entrenamiento o en la selección de variables.  El modelo de árbol de decisión muestra un buen desempeño en la clasificación general, con una exactitud del 79.88%, lo que indica que predice correctamente en la mayoría de los casos. Su alta sensibilidad del 99.61% sugiere que identifica casi todos los admitidos correctamente, minimizando los falsos negativos. Sin embargo, su especificidad es extremadamente baja (2.55%), lo que implica que tiene dificultades para reconocer a quienes no deberían ser admitidos, generando una gran cantidad de falsos positivos. Esto sugiere que el modelo tiende a sobreestimar la admisión, lo que podría afectar la confiabilidad del proceso de selección.